### PR TITLE
CSI-4877 Add vg operator skeleton

### PR DESCRIPTION
**Pull request by ArbelNathan**
Signed-off-by: ArbelNathan <arbel.nathan@ibm.com>

### CODE REVIEW

File: volume-group-operator/main.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		// when the Manager ends. This requires the binary to immediately end when the
+		// Manager is stopped, otherwise, this setting is unsafe. Setting this significantly
+		// speeds up voluntary leader transitions as the new leader don't have to wait
+		// LeaseDuration time first.
+		//
+		// In the default scaffold provided, the program ends immediately after
+		// the manager stops, so would be fine to enable this option. However,
+		// if you are doing or is intended to do any operation such as perform cleanups
+		// after the manager stops then its usage might be unsafe.
+		// LeaderElectionReleaseOnCancel: true,
```
Comment (matancarmeli7) [inline volume-group-operator/main.go:85]: we don't need `HealthProbeBindAddress`, `LeaderElection`, and `LeaderElectionID` parameters, and remove those comments

File: volume-group-operator/main.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	//+kubebuilder:scaffold:builder
+
+	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to set up health check")
+		os.Exit(1)
+	}
+	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to set up ready check")
+		os.Exit(1)
+	}
```
Comment (matancarmeli7) [inline volume-group-operator/main.go:108]: we don't need `healthz` and `readyz` right now, we can remove it

File: volume-group-operator/main.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+func main() {
+	var metricsAddr string
+	var enableLeaderElection bool
+	var probeAddr string
+	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
+	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+	flag.BoolVar(&enableLeaderElection, "leader-elect", false,
+		"Enable leader election for controller manager. "+
+			"Enabling this will ensure there is only one active controller manager.")
```
Comment (matancarmeli7) [inline volume-group-operator/main.go:59]: we don't need `HealthProbeBindAddress`, `LeaderElection`, and `LeaderElectionID` parameters, and not their flags

Comment (ArbelNathan) [reply volume-group-operator/main.go:59]: metricsAddr?

File: volume-group-operator/main.go (hunk 2 above)
Comment (ArbelNathan) [reply volume-group-operator/main.go:108]: maybe we should leave it for now? what is the harm? It sounds important..

File: volume-group-operator/config/crd/kustomizeconfig.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# This file is for teaching kustomize how to substitute name and namespace reference in CRD
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/crd/kustomizeconfig.yaml]: done

File: volume-group-operator/config/crd/patches/cainjection_in_volumegroups.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# The following patch adds a directive for certmanager to inject CA into the CRD
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/crd/patches/cainjection_in_volumegroups.yaml]: done

File: volume-group-operator/config/crd/patches/webhook_in_volumegroups.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# The following patch enables a conversion webhook for the CRD
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/crd/patches/webhook_in_volumegroups.yaml]: done

File: volume-group-operator/config/default/kustomization.yaml [hunk 7]
```diff
@@ -0,0 +1,72 @@
+# Adds namespace to all resources.
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/default/kustomization.yaml]: done

File: volume-group-operator/config/default/manager_auth_proxy_patch.yaml [hunk 8]
```diff
@@ -0,0 +1,55 @@
+# This patch inject a sidecar container which is a HTTP proxy for the
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/default/manager_auth_proxy_patch.yaml]: done

File: volume-group-operator/config/manifests/kustomization.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
+# These resources constitute the fully configured set of manifests
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/manifests/kustomization.yaml]: done

File: volume-group-operator/config/prometheus/kustomization.yaml [hunk 10]
```diff
@@ -0,0 +1,2 @@
+resources:
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/prometheus/kustomization.yaml]: done

File: volume-group-operator/config/rbac/auth_proxy_client_clusterrole.yaml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/rbac/auth_proxy_client_clusterrole.yaml]: done

File: volume-group-operator/config/rbac/auth_proxy_role.yaml [hunk 12]
```diff
@@ -0,0 +1,24 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/rbac/auth_proxy_role.yaml]: done

File: volume-group-operator/config/rbac/leader_election_role.yaml [hunk 13]
```diff
@@ -0,0 +1,44 @@
+# permissions to do leader election.
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/rbac/leader_election_role.yaml]: done

File: volume-group-operator/config/rbac/leader_election_role_binding.yaml [hunk 14]
```diff
@@ -0,0 +1,19 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/rbac/leader_election_role_binding.yaml]: done

File: volume-group-operator/config/rbac/volumegroup_editor_role.yaml [hunk 15]
```diff
@@ -0,0 +1,31 @@
+# permissions for end users to edit volumegroups.
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/rbac/volumegroup_editor_role.yaml]: done

File: volume-group-operator/config/rbac/kustomization.yaml [hunk 16]
```diff
@@ -0,0 +1,18 @@
+resources:
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/rbac/kustomization.yaml:1]: done

File: volume-group-operator/config/rbac/volumegroup_viewer_role.yaml [hunk 17]
```diff
@@ -0,0 +1,27 @@
+# permissions for end users to view volumegroups.
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/rbac/volumegroup_viewer_role.yaml]: done

File: volume-group-operator/config/samples/kustomization.yaml [hunk 18]
```diff
@@ -0,0 +1,4 @@
+## Append samples you want in your CSV to this file as resources ##
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/samples/kustomization.yaml]: done

File: volume-group-operator/config/scorecard/bases/config.yaml [hunk 19]
```diff
@@ -0,0 +1,7 @@
+apiVersion: scorecard.operatorframework.io/v1alpha3
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/scorecard/bases/config.yaml]: done

File: volume-group-operator/config/scorecard/kustomization.yaml [hunk 20]
```diff
@@ -0,0 +1,16 @@
+resources:
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/scorecard/kustomization.yaml]: done

File: volume-group-operator/config/scorecard/patches/basic.config.yaml [hunk 21]
```diff
@@ -0,0 +1,10 @@
+- op: add
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/scorecard/patches/basic.config.yaml]: done

File: volume-group-operator/config/scorecard/patches/olm.config.yaml [hunk 22]
```diff
@@ -0,0 +1,50 @@
+- op: add
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/scorecard/patches/olm.config.yaml]: done

File: volume-group-operator/controllers/volumegroup_controller.go [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the VolumeGroup object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/controllers/volumegroup_controller.go]: done

File: volume-group-operator/go.mod [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+	k8s.io/api v0.25.0 // indirect
+	k8s.io/apiextensions-apiserver v0.25.0 // indirect
+	k8s.io/component-base v0.25.0 // indirect
+	k8s.io/klog/v2 v2.70.1 // indirect
+	k8s.io/kube-openapi v0.0.0-20220803162953-67bda5d908f1 // indirect
+	k8s.io/utils v0.0.0-20220728103510-ee6ede2d64ed // indirect
+	sigs.k8s.io/json v0.0.0-20220713155537-f223a00ba0e2 // indirect
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.3 // indirect
+	sigs.k8s.io/yaml v1.3.0 // indirect
+)
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/go.mod]: done

File: volume-group-operator/main.go (hunk 1 above)
Comment (ArbelNathan) [reply volume-group-operator/main.go]: done

File: volume-group-operator/config/rbac/auth_proxy_role_binding.yaml [hunk 25]
```diff
@@ -0,0 +1,19 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/config/rbac/auth_proxy_role_binding.yaml]: done

File: volume-group-operator/main.go (hunk 2 above)
Comment (matancarmeli7) [reply volume-group-operator/main.go:95]: we don't use it in our operator, if we will want to use it in the future then we will investigate it and add them

File: volume-group-operator/main.go (hunk 3 above)
Comment (matancarmeli7) [reply volume-group-operator/main.go:59]: yes sorry all the 3 parameters there, I copied it from other comment

Comment (ArbelNathan) [reply volume-group-operator/main.go]: done

File: volume-group-operator/main.go (hunk 2 above)
Comment (ArbelNathan) [reply volume-group-operator/main.go:85]: we have a sidecar for this in our operator, it is being used in [replication](https://github.com/csi-addons/volume-replication-operator/blob/b77dac3d3523a0ae446bad6dc1c500647e46aed6/main.go#L151-L158)

File: volume-group-operator/README.md [hunk 26]
```diff
@@ -0,0 +1,105 @@
+# operator
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply volume-group-operator/README.md]: done

File: volume-group-operator/main.go (hunk 2 above)
Comment (matancarmeli7) [reply volume-group-operator/main.go:85]: I don't think we need it, but what you choose

File: volume-group-operator/go.mod [hunk 27]
```diff
@@ -0,0 +1,11 @@
+module github.com/example/memcached-operator
```
Comment (matancarmeli7) [inline volume-group-operator/go.mod:1]: can you change `github.com/example/memcached-operator` to `github.com/IBM/volume-group-operator`? and change all the other appearances of `memcached` accordingly?

File: volume-group-operator/main.go (hunk 2 above)
Comment (ArbelNathan) [reply volume-group-operator/main.go:85]: lets keep it for now

File: volume-group-operator/go.mod (hunk 27 above)
Comment (ArbelNathan) [reply volume-group-operator/go.mod]: done

File: volume-group-operator/config/rbac/role.yaml [hunk 28]
```diff
@@ -0,0 +1,33 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  creationTimestamp: null
+  name: manager-role
```
Comment (matancarmeli7) [inline volume-group-operator/config/rbac/role.yaml:6]: can you change the `ClusterRole`, `ClusterRoleBinding`, and `ServiceAccount` names to volume-group-operator, and adjust accordingly in other related places?

Comment (ArbelNathan) [reply volume-group-operator/config/rbac/role.yaml]: done

**Review by matancarmeli7 — APPROVED** [review]
LGTM

**Pull request merged**